docs(skills): add panel-testing-strategy skill (#129154)
grafana/grafana: .claude/skills/panel-testing-strategy/SKILL.md

---
name: panel-testing-strategy
description: Write unit and E2E tests for Grafana visualization panels and DataViz-owned code the way the DataViz squad wants them. Use when adding or reviewing tests for panels (barchart, timeseries, table, geomap, xychart, heatmap, canvas, stat, etc.), for grafana-ui viz components (Table, uPlot, VizLegend, VizTooltip, DataLinks), or for grafana-data viz utils; when backfilling test coverage; when a panel test only asserts "it rendered" or "it's defined"; when reviewing AI-proposed/AI-generated panel tests for slop; or when a canvas/rendering test is flaky. Encodes the testing-diamond model, assert-real-behavior conventions, the canvas draw-call snapshot harness, no-AI-slop authoring guidance, and anti-flake rules distilled from DataViz PR reviews.
---

# Panel testing strategy

Write tests for Grafana visualization code that reviewers on `@grafana/dataviz-squad` will
accept on the first pass. This encodes the conventions the squad converged on across the

> April–July 2026 coverage effort: **assert concrete behavior, not existence**; keep test

Reviewer: Unnecessary context:

```suggestion
accept on the first pass. Goals: **assert concrete behavior, not existence**; keep test
```

descriptions honest; verify the test actually exercises the target code path; use the
repo's data-frame and panel-props builders; snapshot canvas panels via the draw-call
harness; and stabilize the known flake classes. `@grafana/dataviz-squad` is opted into the
gating `check-frontend-test-coverage.yml` check, so coverage that drops fails CI.

## Resolve the target

Interpret the argument to decide scope:

- **A file path** → test that file (create or extend its co-located `*.test.ts(x)`).
- **A directory / panel name** → the source files under it lacking meaningful coverage.
- **"current file" / no path but a file is open** → the open file.
- **No argument** → ask which panel/area; don't blanket-generate.

Prefer extending an existing co-located test file over adding a new one. Match the
surrounding test file's imports and idiom.

## Principle 1 — Where each test fits: the (inverted) testing diamond

> The squad's model, top to bottom:

Reviewer: ```suggestion
Testing model, top to bottom:
```

- **E2E** (pinnacle) — validate the system via real user flows; powerful but slow, so keep
  it targeted and few.

> - **Visual regression** (middle) — catch UI / interaction drift (handled by Meticulous).

Reviewer: If not relevant to the task of writing tests I say remove it

  and cover nothing. Set every precondition the branch requires.
- **Assert collaboration precisely** when you deliberately don't want to test a collaborator —
  mock it and verify it's called with exact args / counts, not just that output exists:

```ts
const guess = jest.spyOn(mod, 'guessFieldTypes');
processFrames([frameA, frameB]);
expect(guess).toHaveBeenCalledTimes(2);              // once per frame — use ≥2 frames
expect(getColor.mock.calls.map((c) => c[1])).toEqual([0, 2]); // skipped null at idx 1
guess.mockRestore();
```

Avoid loose assertions that pass for the wrong reason: no `toMatch(/50/)` where the exact
value is knowable; no `toBeGreaterThanOrEqual` where the code guarantees a strict change
(use `toBeGreaterThan`). Confine any external-interface cast to one helper
(`asUPlot(u)`, `as unknown as DataFrameDTO`) rather than sprinkling `@ts-expect-error`.

## Step 4 — Canvas / rendering panels: use the draw-call snapshot harness

Panels that draw to canvas (timeseries, heatmap, xychart, timeline, piechart, sparkline)
are tested by **capturing the ctx draw-call stream**, not by pixel-diffing (visual

> regression is Meticulous's job). Follow the established harness:

Reviewer: ```suggestion
are tested by **capturing the ctx draw-call stream**, not by pixel-diffing. Follow the established harness:
```

```ts
import { createGrafanaUiMeasureTextJestMock, installCanvasPath2DShim,
         removeCanvasTransforms } from '@grafana/test-utils/canvas';

> // see public/app/plugins/panel/timeseries/TimeSeriesPanel.canvasTestUtils.tsx

Reviewer: TimeSeriesPanel.canvasTestUtils.tsx actually imports `applyDefaultUPlotAxisMeasureTextMock`, might be worth taking another look here to align

Author: should be fixed by https://github.com/grafana/grafana/pull/129154/commits/3f13a700210555769a84420fb6cfdbbefcee070d - when fixing this, the agent recommended the cleanup here too, so I went ahead and opened it. https://github.com/grafana/grafana/pull/129467

per panel type: default render, **hover / tooltip open**, **legend item toggled**, **sort /
filter applied** (table), **series selected**, **panel edit mode**, and **empty / no-data**.

```ts
const panel = dashboardPage.getByGrafanaSelector(selectors.components.Panels.Panel.content);
await expect(panel).toMatchAriaSnapshot();               // baseline structure
await panel.locator('.uplot').hover({ position: { x: 120, y: 80 } });
await expect(panel).toMatchAriaSnapshot();               // tooltip-open state
expect(await scanForA11yViolations()).toHaveNoA11yViolations(); // a11y holds mid-interaction
```

Keep these deterministic — see the anti-flake rules below (pin data, scope locators, wait
for the renderer). Aria snapshots capture semantic structure, not pixels; leave pixel-level
visual regression to Meticulous.

## Step 6 — Don't test what shouldn't exist

- **Skip modules slated for deletion.** Adding tests to deprecated code (e.g. the Vector
  classes) signals it's load-bearing and obstructs its removal. If unsure, ask.
- Deferring comprehensiveness to a follow-up PR is acceptable — leave an explicit note
  rather than shipping a shallow test that looks complete.

> ## Anti-flake rules (from tests disabled & re-enabled, Apr–Jul 2026)

Reviewer: ```suggestion
## Anti-flake rules
```

Each rule maps to a real DataViz stabilization; global Playwright config retries once in CI
only. **Avoid → Do:**

1. **Non-deterministic data / relative time ranges.** Avoid `random_walk` + `now-30m` when
   asserting on shapes or coordinates. Do pin an absolute time range and a fixed seed /
   `startValue` so the render is identical every run. _(timeseries tooltip #128617)_
2. **Coordinate-based hover/click on canvas.** Avoid hardcoded x/y on `.uplot`. Do make data
   deterministic first; derive coords from rendered geometry, not constants. _(xychart tooltip
   remains skipped #128389 for this reason)_
3. **Broad locators.** Avoid `page.locator('.uplot')` — it also matches option-pane preview
   thumbnails. Do scope: `getByGrafanaSelector(Panels.Panel.content).locator('.uplot')`.

> 4. **Asserting before the renderer is ready.** Do `await waitFor(() => uPlotInstance?.status === 1)`

Reviewer: waitFor without expect just resolves without throwing:

https://testing-library.com/docs/dom-testing-library/api-async/#waitfor
>Returning a falsy condition is not sufficient to trigger a retry, the callback must throw an error in order to retry the condition.
```suggestion
4. **Asserting before the renderer is ready.** Do `waitFor(() => expect(uPlotInstance?.status).toBe(1))`
```

Reviewer: We've got a few of the bad checks in use that we probably want to clean up as well (`HeatmapPanel.canvas.test.tsx`, `SparklineCell.canvas.test.tsx`, `XYChartPanel.canvas.test.tsx`)

- E2E for interaction; define selectors in `@grafana/e2e-selectors` first; query by selector.
- Every panel gets an `@a11y` E2E test (`scanForA11yViolations` + `toHaveNoA11yViolations`),
  plus interaction snapshots (`toMatchAriaSnapshot`) across a variety of states
  (hover/tooltip, legend toggle, sort/filter, edit, empty) — not just first render.
- Don't add tests to code slated for deletion.
- Apply the 10 anti-flake rules; never assert on a canvas before `status === 1`.
- Run the codeowner coverage check before pushing (it gates the PR).
- Never commit an unfocused, verbose, or implementation-coupled "slop" test; no unreadable
  DOM snapshot tests. Review AI output thoroughly before opening a PR.
- Tests land in the same PR as the feature or bug fix; aim for intentional coverage, not
  incidental, and close every bug with a regression test at the right layer.

## Exemplar files

- Behavior-specific util tests with typed uPlot mocks & `it.each`:
  `public/app/plugins/panel/barchart/bars.test.ts`
- Spy-per-frame, untyped-input inference, distinct loading cases:
  `packages/grafana-data/src/dataframe/processDataFrame.test.ts`
- Concrete-value assertions & clear descriptions:

>   `packages/grafana-ui/src/components/Table/{utils,cellUtils}.test.ts`,

Reviewer: there is no cellUtils.test.ts

Author: https://github.com/grafana/grafana/blob/main/packages/grafana-ui/src/components/Table/cellUtils.test.ts ?

Reviewer: Oh weird I'm guessing my IDE index was stale. NVM :homer-disappear

>   `packages/grafana-ui/src/components/uPlot/config/gradientFills.test.ts`

Reviewer: `gradientFills.test.ts` doesn't exist either

Author: https://github.com/grafana/grafana/blob/main/packages/grafana-ui/src/components/uPlot/config/gradientFills.test.ts ?

Author: not saying there aren't better examples!

Reviewer: sorry, plz ignore

---
name: panel-testing-strategy
description: Write unit and E2E tests for Grafana visualization panels and viz utilities to the conventions this repo expects. Use when adding, backfilling, or reviewing tests for panels (barchart, timeseries, table, xychart, heatmap, canvas, etc.), grafana-ui viz components (Table, uPlot, VizLegend, VizTooltip), or grafana-data viz utils; when a panel test only asserts "it rendered" or "it's defined"; when reviewing AI-generated panel tests for slop; or when a canvas/rendering test is flaky.
---

# Panel testing strategy

Write tests for Grafana visualization code that pass review on the first pass. Goals:
**assert concrete behavior, not existence**; keep test descriptions honest; verify the test
actually exercises the target code path; use the repo's data-frame and panel-props builders;

> snapshot canvas panels via the draw-call harness; and stabilize the known flake classes. The

Reviewer 2: **nit(non-blocking):** disambiguate

```suggestion
snapshot HTML5 canvas based panels via the draw-call harness; and stabilize the known flake classes. The
```

derives from) and confirm the test goes **red**. A test that stays green — because its expectation
tracks the code, or checks a value against itself — is a tautology dressed as coverage. Make this a
habit, not just the final Verify step.

## Principle 3 — Authoring with AI: no slop tests

This skill exists so AI-proposed tests meet the bar above. The failure mode to avoid is the
**slop test**:

- **Unfocused** — a wide blast of assertions that doesn't preserve the intent of the code
  under test.
- **Verbose** — unnecessary steps/mocks for a simple goal; brittle to implementation
  changes, and can silently mask real regressions.
- **Limiting** — so many, or so coupled to implementation, that a later refactor breaks them
  without telling you whether behavior actually broke. (Unreadable DOM snapshot tests are the
  classic example — never add them.)

Review AI output _thoroughly_ before opening a PR; expect to amend it for
readability/maintainability. If reviewing the AI output costs more than writing the test by
hand, write it by hand. A test is a specification a teammate — and future-you — must read
easily; value refactoring for readability over a raw coverage percentage.

> 

Reviewer 2: **suggestion(non-blocking):** This can be a foot gun with AI generated tests, and I didn't notice it mentioned elsewhere.

```suggestion

## Principle 4 - Do not simply update failing tests to pass after changing behaviour, or adding a feature

When a test fails after updating functionality, behaviour or features, this is a warning that a regression was caused. Tests are meant as a safety net to catch regressions, and a failing test isn't broken -- its doing its job.

Net new functionality requires net new tests. Existing tests should be treated as a spec, and if you cause a test to fail -- STOP -- and analyze why that test fails, only after thorough analysis based on tracing real code should an existing test ever be updated.

```

Author: excellent addition

> ## Step 1 — Set up data with the repo's builders

Reviewer 2: **suggestion(non-blocking):" Another common AI generated test footgun, take-it-or-leave-it.

```suggestion
## Principle 5 - Tests are meant to be read by human eyes, be concise

Favour self-documenting code, semantically meaningful variable names, and sensible test descriptions. Avoid inline code comments, as much as possible by writing self-documenting tests. Never write multi-line inline code comments for tests.

Keep it short and sweet.

When possible, use table tests to improve the signal to noise ratio for future readers of the tests you write.

## Step 1 — Set up data with the repo's builders
```

Author: The tests that have been produced by this skill have had some comment blocks, but they have been in service of explaining some of the more dense parts of the code, crufty test code that may not be possible to write more elegantly, or something that just may be confusing to someone reading the test, so I worry that the word "avoid" re: comments would cause the agents to omit some of the useful ones they've been adding.

For example, this [geomap PR](https://github.com/grafana/grafana/pull/129762) went through the skill and I think the comments it added are in the sweet spot for me where they are all pretty helpful in enriching my understanding of the tests.

I'd be ok with taking the principle but softening the language around comments, wdyt?

> Favour self-documenting code, semantically meaningful variable names, and sensible test descriptions. When possible, the test code should be self-evident, and code comments should be reserved for situations where it's not possible. Try to avoid multi-line inline code comments within tests.

Reviewer 2: Totally, if you think this is sufficiently covered already, I'm OK with whatever.

For me the pattern has generally been LLMs running comment-heavy, which obvs leads to comment-rot over time.

Non-blocking, take-it-or-leave-it feedback, I'm OK with whatever you decide.

processFrames([frameA, frameB]);
expect(guess).toHaveBeenCalledTimes(2); // once per frame — use ≥2 frames
expect(getColor.mock.calls.map((c) => c[1])).toEqual([0, 2]); // skipped null at idx 1
guess.mockRestore();
```

Avoid loose assertions that pass for the wrong reason: no `toMatch(/50/)` where the exact
value is knowable; no `toBeGreaterThanOrEqual` where the code guarantees a strict change
(use `toBeGreaterThan`). Confine any external-interface cast to one helper
(`asUPlot(u)`, `as unknown as DataFrameDTO`) rather than sprinkling `@ts-expect-error`.

To type a mocked function or module, use `jest.mocked(fn)` — never
`fn as jest.MockedFunction<typeof fn>` (or `as jest.Mocked<…>`). `jest.mocked` is the
type-safe, less noisy repo convention and gives typed access to `.mock` / `.mockReturnValue`:

```ts
import { measureText } from '@grafana/ui';
jest.mock('@grafana/ui', () => ({ ...jest.requireActual('@grafana/ui'), measureText: jest.fn() }));

const measureTextMock = jest.mocked(measureText); // ✅ not `measureText as jest.MockedFunction<…>`
measureTextMock.mockReturnValue({ width: 100 } as TextMetrics);
expect(measureTextMock).toHaveBeenCalledWith('label', 12);
```

> ## Step 4 — Canvas / rendering panels: use the draw-call snapshot harness

Reviewer 2: **nit(non-blocking):** Disambiguate

```suggestion
## Step 4 — HTML5 canvas / rendering panels: use the draw-call snapshot harness
```
